uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php

```diff
@@ -135,4 +145,238 @@ public function testDeleteDeletesTheLeaveRequestItsBalanceChangesAndDates() {
 
     $this->fail("Expected to not find the LeaveRequest with {$leaveRequest->id}, but it was found");
   }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You are not allowed to create or update a leave request for this employee
+   */
+  public function testCreateThrowsAnExceptionWhenCurrentUserDoesNotHaveCreateAndUpdateLeaveRequestPermission() {
+    //logged in user has no permissions, also a contactID different from that of the logged in user is passed
+    $contactID = 2;
+    $this->getleaveRequestService()->create([
+      'type_id' => 1,
+      'contact_id' => $contactID,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You can't create a Leave Request with this status
+   */
+  public function testCreateThrowsAnExceptionWhenTransitionStatusIsNotValidForNewLeaveRequest() {
+    //A leave contact trying to set a new leave request to approved status
+    $this->getLeaveRequestServiceWhenStatusTransitionIsNotAllowed()->create([
+      'type_id' => 1,
+      'contact_id' => $this->leaveContact,
+      'status_id' => $this->leaveRequestStatuses['Approved']['id'],
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
```

Comment:
Are you sure they're different for most of the tests? I just counted 11 tests all using the same dates, date types, type, contact etc. The main difference seems to be only the status.

If you need some different params for some tests, you can do something like what we do for fabricators:

```php
public function getDefaultParams($params) {
  $defaultParams = [/*list of default params*/];
   return array_merge($default_params, $params);
}
```

Then, if you need a different status for a test, you can do:
```php
$params = $this->getDefaultParams([
  'status_id' => $statuses['Cancelled']['value']
]);
```

The idea is to avoid polluting the test with information/data that is not really relevant to it.